modules/fedimint-mint-client/src/lib.rs

```diff
@@ -377,29 +392,88 @@ impl MintClientModule {
     }
 }
 
-// TODO: remove once streaming selection is implemented
-pub fn select_notes<C: Clone>(notes: TieredMulti<C>, mut amount: Amount) -> Option<TieredMulti<C>> {
-    if amount > notes.total_amount() {
-        return None;
+// We are using a greedy algorithm to select notes. We start with the largest
+// then proceed to the lowest tiers/denominations.
+// But there is a catch: we don't know if there are enough notes in the lowest
+// tiers, so we need to save a big note in case the sum of the following
+// small notes are not enough.
+pub async fn select_notes_from_stream<Note>(
+    stream: impl futures::Stream<Item = (Amount, Note)>,
+    requested_amount: Amount,
+) -> Result<TieredMulti<Note>, InsufficientBalanceError> {
+    if requested_amount == Amount::ZERO {
+        return Ok(TieredMulti::default());
     }
-
+    let mut stream = Box::pin(stream);
     let mut selected = vec![];
-    let mut remaining = notes.total_amount();
-
-    for (note_amount, note) in notes.iter_items().rev() {
-        remaining -= note_amount;
-
-        if note_amount <= amount {
-            amount -= note_amount;
-            selected.push((note_amount, (*note).clone()))
-        } else if remaining < amount {
-            // we can't make exact change, so just use this note
-            selected.push((note_amount, (*note).clone()));
-            break;
+    // This is the big note we save in case the sum of the following small notes are
+    // not sufficient to cover the pending amount
+    // The tuple is (amount, note, checkpoint), where checkpoint is the index where
+    // the note should be inserted on the selected vector if it is needed
+    let mut last_big_note_checkpoint: Option<(Amount, Note, usize)> = None;
+    let mut pending_amount = requested_amount;
+    let mut previous_amount: Option<Amount> = None; // used to assert descending order
+    loop {
+        if let Some((note_amount, note)) = stream.next().await {
+            assert!(
+                previous_amount.map_or(true, |previous| previous >= note_amount),
```

Comment:
Wonder how much this actually saves in practice...I'd suspect in most cases you'd need to load every note down to 1sat in order to make exact change.